### PR TITLE
EOS-17389: LR R2: Mini Provisioner - Convert s3_setup to python - config phase <patch 2: Convert config phase from shell to python>

### PR DESCRIPTION
**Pull request by pratyush-seagate**
commit 97d8d582068ac66626da9b720b100e9a113a2825
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Thu Feb 18 00:29:53 2021 -0700

        EOS-17389: LR R2: Mini Provisioner - Convert s3_setup to python - config phase <patch 2: Convert config phase from shell to python>
        Change description:
                As described in ticket EOS-17389 for converting the shell steps to python for config phase
        Added/modified/deleted files:
                modified:   scripts/provisioning/configcmd.py
                modified:   scripts/provisioning/s3_prov_config.yaml
                modified:   scripts/provisioning/s3_setup
                modified:   scripts/provisioning/setupcmd.py
            Unit test:
                    Run "s3_setup config" with these changes and verified steps worked. Performed complete mini-provisioner s3 steps

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

### CODE REVIEW

File: scripts/provisioning/configcmd.py [hunk 1]
```diff
@@ -30,11 +33,119 @@ def __init__(self, config: str):
...
     """Main processing function."""
     retval = 0
     sys.stdout.write(f"Processing {self.name} {self.url}\n")
+
+    # Configure openldap and ldap-replication
+    try:
+      retval = self.configure_openldap()
+    except Exception as e:
+      sys.stderr.write(f'In self.configure_openldap(), exception: {e}\n')
+      return 1
```
Comment (amitwac2608) [inline scripts/provisioning/configcmd.py:61]: no need of returning any value, raise the exception up

File: scripts/provisioning/configcmd.py [hunk 2]
```diff
@@ -30,11 +33,119 @@ def __init__(self, config: str):
...
+    if retcode != 0:
+      sys.stderr.write(f"{cmd} failed with err: {stderr}, out: {stdout}, ret: {retcode}\n")
+      return 1
+    os.remove("hosts_list_file.txt")
+
+    return 0
+
+  def configure_haproxy(self):
+    """Configure haproxy service."""
+    cmd = ['s3haproxyconfig',
```
Comment (amitwac2608) [inline scripts/provisioning/configcmd.py:142]: We should not call s3haproxyconfig binary, instead us haproxy class package. S3Haproxyconfig is also in python.

File: scripts/provisioning/configcmd.py [hunk 3]
```diff
@@ -30,11 +33,119 @@ def __init__(self, config: str):
...
+      return 1
+
+    if retval == 0:
+      sys.stdout.write("INFO: Successfully configured openldap on the node.\n")
+      # Configure haproxy
+      try:
+        retval = self.configure_haproxy()
+      except Exception as e:
+        sys.stderr.write(f'In self.configure_haproxy(), exception: {e}\n')
+        return 1
```
Comment (amitwac2608) [inline scripts/provisioning/configcmd.py:70]: do not return anything, just raise the exception up

File: scripts/provisioning/configcmd.py [hunk 4]
```diff
@@ -30,11 +33,119 @@ def __init__(self, config: str):
...
+    # 3. Set openldap-replication
+    # 4. Check number of nodes in the cluster
+    cmd = ['/opt/seagate/cortx/s3/install/ldap/setup_ldap.sh',
+           '--ldapadminpasswd',
+           f'{self.ldap_passwd}',
+           '--rootdnpasswd', 
+           f'{self.rootdn_passwd}',
+           '--skipssl']
+    handler = SimpleProcess(cmd)
+    stdout, stderr, retcode = handler.run()
```
Comment (amitwac2608) [inline scripts/provisioning/configcmd.py:95]: I would suggest to use exceptions instead of return values. refer SubProcess.run() documentation :
https://docs.python.org/3/library/subprocess.html

_If check is true, and the process exits with a non-zero exit code, a CalledProcessError exception will be raised. Attributes of that exception hold the arguments, the exit code, and stdout and stderr if they were captured._

Comment (pratyush-seagate) [reply scripts/provisioning/configcmd.py:95]: To address this, I'll need to add changes in cortx-utils repo. If instead I try to directly use SubProcess() APIs in this code, codacy will raise warnings as it is against using subprocess(). Hence pls let me know if it is ok to go with this change, otherwise I'll need to address this in cortx-utils repo.

Comment (amitwac2608) [reply scripts/provisioning/configcmd.py:95]: ohh sorry.. I confused cortx-utils.SimpleProcess with SubProcess.

I guess we need to live with return codes, until we change the setup_ldap.sh to python also.

Comment (pratyush-seagate) [reply scripts/provisioning/configcmd.py:76]: As discussed, used custom exception to trigger in such failure cases. Also tested negative situations using manual fault injection in code and verified they are working.

File: scripts/provisioning/configcmd.py (hunk 2 above)
Comment (pratyush-seagate) [reply scripts/provisioning/configcmd.py:114]: Following the discussions with @amitwac2608  and @shahidSTX , it seems we need to update ./scripts/haproxy/s3haproxyconfig/s3haproxyconfig/s3_haproxy_config.py and remove it from an executable, and restructure the current code and put it into a new file under scripts/provisioning/. The restructuring will also address modularization of the existing large run() function in s3_haproxy_config.py. This significantly affects the scope of work of the current ticket.

@AjinkyaDhumal  please suggest if this needs to be done as part of the current config conversion ticket https://jts.seagate.com/browse/EOS-17389 or can be addressed in a separate ticket later after EOS-17389 is resolved.

**Comment by pratyush-seagate** [comment]
image

this issue: D203 is in conflict with D211, which says there should not be any blank line. I found this as existing issue :

https://github.com/PyCQA/pydocstyle/issues/141

We should also disable D203 for our project.

**Comment by pratyush-seagate** [comment]
Note: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/321/console
mini-provisioner job is failing in config phase due to failure in /opt/seagate/cortx/s3/install/ldap/setup_ldap.sh. Debugging.

File: scripts/provisioning/configcmd.py (hunk 2 above)
Comment (pratyush-seagate) [reply scripts/provisioning/configcmd.py:114]: Note: New ticket https://jts.seagate.com/browse/EOS-18174 created to track this task in the next sprints

**Comment by pratyush-seagate** [comment]
> 
> 
> Note: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/321/console
> mini-provisioner job is failing in config phase due to failure in /opt/seagate/cortx/s3/install/ldap/setup_ldap.sh. Debugging.

This failure was due to missing call create_auth_jks_password() during config phase. It seems we must need to call create_auth_jks_password() before executing setup_ldap.sh().

Since @sachitanands change for create_auth_jks_password() is not available yet, I didn't remove the call to  create_auth_jks_password() from s3_setup (shell version). Please review @AjinkyaDhumal  as this is different from what is mentioned in https://jts.seagate.com/browse/EOS-17389's DoD.

**Comment by pratyush-seagate** [comment]
> 
> 
> > Note: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/321/console
> > mini-provisioner job is failing in config phase due to failure in /opt/seagate/cortx/s3/install/ldap/setup_ldap.sh. Debugging.
> 
> This failure was due to missing call create_auth_jks_password() during config phase. It seems we must need to call create_auth_jks_password() before executing setup_ldap.sh().
> 
> Since @sachitanands change for create_auth_jks_password() is not available yet, I didn't remove the call to create_auth_jks_password() from s3_setup (shell version). Please review @AjinkyaDhumal as this is different from what is mentioned in https://jts.seagate.com/browse/EOS-17389's DoD.

@sachitanands  informed that there is no need as of now to remove the create_auth_jks_password() as there is no overlapping work. So this point is good now and this PR can be taken forward.

File: scripts/provisioning/setupcmd.py [hunk 5]
```diff
@@ -73,3 +85,27 @@ def read_ldap_credentials(self):
...
       sys.stderr.write(f'unknown exception: {e}\n')
       raise e
+
+  def read_cluster_id(self):
+    """Get 'cluster>cluster_id' from confstore."""
+    prov_config_yaml = None
+
+    try:
+      with open(self.s3_prov_config) as provconf:
+        prov_config_yaml = yaml.safe_load(provconf)
```
Comment (amitwac2608) [inline scripts/provisioning/setupcmd.py:95]: We have been advised to use confstore to load and read any config file. refer to my PR: https://github.com/Seagate/cortx-s3server/pull/707

File: scripts/provisioning/s3_setup [hunk 6]
```diff
@@ -267,7 +220,14 @@ create_background_delete_account()
 if [ $s3postinstall == true ]
 then
     echo "INFO: Start haproxy, slapd, rabbitmq-server and rsyslog services.."
-    systemctl restart haproxy slapd rabbitmq-server rsyslog
+    sh /opt/seagate/cortx/s3/bin/s3prov_start_services.sh haproxy slapd rabbitmq-server rsyslog
+    if [ $? -eq 0 ]
+    then
+      echo "Services are running successfully."
+    else
+      echo "Failed to start services."
```
Comment (amitwac2608) [inline scripts/provisioning/s3_setup:228]: I am not sure, but can we give info on which service failed to start !!!

Comment (pratyush-seagate) [reply scripts/provisioning/s3_setup:228]: Yes, /opt/seagate/cortx/s3/bin/s3prov_start_services.sh already prints this information internally.

**Review by shahidSTX — APPROVED** [review]
LGTM.

File: scripts/provisioning/setupcmd.py (hunk 5 above)
Comment (pratyush-seagate) [reply scripts/provisioning/setupcmd.py]: thanks addressed this point.

File: scripts/provisioning/s3_setup [hunk 7]
```diff
@@ -281,31 +241,18 @@ then
...
-    fetch_update_cluster_id
-
-    # get openldap password from confstore
-    echo "INFO: Get openldap passwords from confstore.."
-    get_ldap_password
 
+    # NOTE: Call to create_auth_jks_password will be removed later
     # Create random password for auth jks keystore
     echo "INFO: Creating random jks password for Auth server.."
     create_auth_jks_password
```
Comment (nileshgovande) [inline scripts/provisioning/s3_setup:248]: ToBe evaluated

Comment (nileshgovande) [reply scripts/provisioning/s3_setup:248]: Create Cleanup JIRA

File: scripts/provisioning/s3_setup [hunk 8]
```diff
@@ -281,31 +241,18 @@ then
     # validate pre-requisites before proceeding
     # we already have 'exit' statement in '_s3_setup' binary at failure scenario,
     # no need to add 'exit' here.
-    echo "INFO: validating pre-requisites for provisioning s3server.."
-    $install_dir/_s3_setup post_install
-
-    # Fetch and update cluster id
-    # To decrypt openldap passwords, we need 'cluster_id' updated in s3_cluster.yaml
-    echo "INFO: Fetching and updating cluster id in s3_cluster.yaml .."
-    fetch_update_cluster_id
```
Comment (nileshgovande) [inline scripts/provisioning/s3_setup:290]: Remove function definition as well

File: scripts/provisioning/setupcmd.py [hunk 9]
```diff
@@ -51,25 +57,38 @@ def s3confstore(self) -> str:
...
+  def read_cluster_id(self):
+    """Get 'cluster>cluster_id' from confstore."""
+
+    try:
+      localconfstore = S3CortxConfStore(f'yaml://{self.s3_prov_config}', 'read_cluster_ididx')
+      self.cluster_id = self.s3confstore.get_config(localconfstore.get_config('CONFSTORE_CLUSTER_ID_KEY'))
+    except Exception as e:
+      raise S3PROVError(f'exception: {e}\n')
+
+  def read_node_info(self):
```
Comment (nileshgovande) [inline scripts/provisioning/setupcmd.py:88]: Need to write clusterID file

File: scripts/provisioning/s3_setup (hunk 8 above)
Comment (pratyush-seagate) [reply scripts/provisioning/s3_setup:290]: done

File: scripts/provisioning/setupcmd.py (hunk 9 above)
Comment (pratyush-seagate) [reply scripts/provisioning/setupcmd.py:98]: fixed, thanks for catching this

File: scripts/provisioning/s3_setup (hunk 7 above)
Comment (pratyush-seagate) [reply scripts/provisioning/s3_setup:234]: Should I create this ticket ?

**Comment by pratyush-seagate** [comment]
@nileshgovande  @AjinkyaDhumal 
PR https://github.com/Seagate/cortx-s3server/pull/705 all reviews are addressed. S3 premerge mini-provisioning also passed, jenkins runnins. @amitwac2608  triggered http://eos-jenkins.mero.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-R2-1Node/74/ on custom build http://cortx-storage.colo.seagate.com/releases/cortx/github/integration-custom-ci/release/centos-7.8.2003/custom-build-669/cortx_iso/, I'll monitor this job.

File: scripts/provisioning/s3_setup (hunk 7 above)
Comment (pratyush-seagate) [reply scripts/provisioning/s3_setup:234]: @sachitanands  please create a ticket for this.

Comment (sachitanands) [reply scripts/provisioning/s3_setup:234]: @pratyush-seagate , @nileshgovande created ticket : https://jts.seagate.com/browse/EOS-18331 for this JKS related work

File: scripts/provisioning/setupcmd.py [hunk 10]
```diff
@@ -51,25 +57,48 @@ def s3confstore(self) -> str:
...
+      self.cluster_id = self.s3confstore.get_config(localconfstore.get_config('CONFSTORE_CLUSTER_ID_KEY'))
+    except Exception as e:
+      raise S3PROVError(f'exception: {e}\n')
+
+  def write_cluster_id(self, op_file: str = "/opt/seagate/cortx/s3/s3backgrounddelete/s3_cluster.yaml"):
+    """Set 'cluster>cluster_id' to op_file."""
+
+    try:
+      with open(f'{op_file}', 'w+') as fhndle:
+        fhndle.write("cluster_config:\n")
```
Comment (amitwac2608) [inline scripts/provisioning/setupcmd.py:93]: this will override everything in the file, including the copyright header.  Try to see how to do "sed" in python.

Comment (pratyush-seagate) [reply scripts/provisioning/setupcmd.py]: Addressed this using confstore to write the value. Please review the latest changes.

Comment (pratyush-seagate) [reply scripts/provisioning/setupcmd.py]: FYI:
confstore save API is also removing the comment out copyright info. Created https://jts.seagate.com/browse/EOS-18356 to track this issue as discussed.

**Review by nileshgovande — APPROVED** [review]
